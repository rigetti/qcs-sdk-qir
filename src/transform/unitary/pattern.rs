// Copyright 2022 Rigetti Computing
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::{collections::HashMap, sync::LazyLock};

use eyre::{eyre, Result};
use inkwell::{
    basic_block::BasicBlock,
    values::{FloatValue, InstructionOpcode, InstructionValue},
};
use log::{debug, info};
use qcs::quil_rs::instruction::{Gate, Instruction, Measurement};
use qcs::quil_rs::Program;
use qcs::quil_rs::{
    expression::Expression,
    instruction::{GateModifier, MemoryReference, Qubit},
};
use regex::Regex;

use crate::{
    context::QCSCompilerContext,
    interop::instruction::{
        get_called_function_name, get_qis_function_arguments, OperationArgument,
    },
    transform::PARAMETER_MEMORY_REGION_NAME,
    RecordedOutput,
};

/// Each pattern matching function, if it finds a match, returns a result tuple including:
/// * data about the pattern which was matched.
/// * what the next instruction following the pattern is, if any.
type PatternResult<'ctx, T> = Option<(Option<InstructionValue<'ctx>>, T)>;

/// A `UnitaryPatternMatchContext` accumulates state as it scans a QIR program.
/// It is used to extract a single quantum program from a QIR program, without shot count
/// or any classical instructions beyond a final return statement.
///
/// The pattern it looks for within a basic block is the following:
///
/// * A sequence of these instructions only:
///   * Quantum Instructions - `quantum_qis` intrinsics
///   * Select Quantum Runtime Instructions - `quantum_rt`
/// * Terminated by _exactly_ a `ret void`
///
/// If any other instructions are encountered, an error is returned.
#[derive(Debug, Default)]
pub(crate) struct UnitaryPatternMatchContext<'ctx> {
    /// The quil program transpiled from quantum intrinsics
    pub(crate) quil_program: Program,

    /// Signifies output to be recorded at the end of program execution
    pub(crate) recorded_output: Vec<RecordedOutput>,

    /// A list of instructions to remove from the program (for substitution with Quil)
    pub(crate) instructions_to_remove: Vec<InstructionValue<'ctx>>,

    /// Mapping of `(read_result *Result index) - >(ro memory region index)`
    pub(crate) read_result_mapping: HashMap<u64, u64>,

    /// Pairings of (readout buffer index/offset) with the instruction which stores that readout value.
    pub(crate) readout_instruction_mapping: Vec<(u64, InstructionValue<'ctx>)>,

    /// All `FloatValues` used as instruction parameters. Indices within this Vec map to indices within the Quil
    /// `MemoryRegion` used to read the values at runtime.
    pub(crate) parameters: Vec<FloatValue<'ctx>>,

    /// Whether or not to prepend a RESET instruction to the program to actively reset all qubits on each shot
    pub(crate) use_active_reset: bool,
}

impl<'ctx> UnitaryPatternMatchContext<'ctx> {
    /// Build the pattern context from a basic block.
    ///
    /// # Arguments
    ///
    /// * `context`: overall compiler context
    /// * `basic_block`: the subject block to be searched for the pattern
    pub(crate) fn from_basic_block(
        context: &mut QCSCompilerContext<'ctx>,
        basic_block: BasicBlock<'ctx>,
    ) -> Result<Self> {
        let mut next_instruction = basic_block.get_first_instruction();
        let mut pattern_context = UnitaryPatternMatchContext::default();

        info!(
            "starting transpile: block {}",
            basic_block.get_name().to_str()?
        );

        while let Some(instruction) = next_instruction {
            // If we haven't yet found the loop start...
            if let Some((pattern_instruction, ())) =
                quantum_instruction(context, &mut pattern_context, instruction)?
            {
                debug!("matched quantum instruction: {instruction:?}");
                next_instruction = pattern_instruction;
            } else if let Some((pattern_instruction, ())) =
                rt_record_instruction(context, &mut pattern_context, instruction)?
            {
                debug!("matched rt_record instruction: {instruction:?}");
                next_instruction = pattern_instruction;
            } else if instruction.get_opcode() == InstructionOpcode::Return {
                return Ok(pattern_context);
            } else {
                return Err(eyre::eyre!(
                    "found instruction disallowed in Unitary QIR: {:?}",
                    instruction
                ));
            }
        }

        Ok(pattern_context)
    }

    /// If the program contains any executable instructions (gates, pulses, etc) return that
    /// information; otherwise, return `None` indicating that the pattern was not matched.
    pub(crate) fn get_program_data(&self) -> Option<&Program> {
        if self.quil_program.body_instructions().count() == 0 {
            None
        } else {
            Some(&self.quil_program)
        }
    }

    /// Returns the parameters which do not have a constant value.
    pub(crate) fn get_dynamic_parameters(&self) -> Vec<&FloatValue<'ctx>> {
        self.parameters
            .iter()
            .filter(|v| !v.is_const())
            .collect::<Vec<&FloatValue>>()
    }
}

macro_rules! match_qis_argument {
    ($variant:ident, $arguments:expr, $index:expr, $function_name:expr) => {{
        use OperationArgument::*;

        match $arguments.get($index) {
            Some($variant(contents)) => Ok(contents),
            other => Err(eyre!(
                "expected argument {} of {} to be of type {:?}; got {:?}",
                $index,
                $function_name,
                stringify!($variant),
                other
            )),
        }
    }};
}

/// Given a `FloatValue` which may be the parameter of a QIS intrinsic call, return the Quil Expression
/// which will be used to store its value at execution time.
fn get_quil_parameter_expression<'ctx>(
    pattern_context: &mut UnitaryPatternMatchContext<'ctx>,
    float_value: FloatValue<'ctx>,
) -> Expression {
    if let Some((constant, _)) = float_value.get_constant() {
        return Expression::Number(constant.into());
    }

    let index = get_quil_parameter_index(pattern_context, float_value);
    Expression::Address(MemoryReference {
        name: String::from(PARAMETER_MEMORY_REGION_NAME),
        index: index as u64,
    })
}

/// Given a `FloatValue` to be used as the parameter to a gate, return the index within the
/// Quil `MemoryReference` that should be used to store this parameter's value.
pub(crate) fn get_quil_parameter_index<'ctx>(
    pattern_context: &mut UnitaryPatternMatchContext<'ctx>,
    float_value: FloatValue<'ctx>,
) -> usize {
    if let Some(index) = pattern_context
        .parameters
        .iter()
        .position(|el| el == &float_value)
    {
        index
    } else {
        pattern_context.parameters.push(float_value);
        pattern_context.parameters.len() - 1
    }
}

#[allow(clippy::too_many_arguments)]
fn add_gate_instruction<'ctx>(
    pattern_context: &mut UnitaryPatternMatchContext<'ctx>,
    arguments: &[OperationArgument<'ctx>],
    function_name: &str,
    name: &str,
    adjoint: bool,
    controlled: bool,
    parameter_count: usize,
    qubit_count: usize,
) -> Result<()> {
    let parameters = (0..parameter_count)
        .map(|arg_index| {
            let float_value = *match_qis_argument!(Parameter, arguments, arg_index, function_name)?;
            Ok(get_quil_parameter_expression(pattern_context, float_value))
        })
        .collect::<Result<Vec<Expression>>>()?;

    let qubits = (parameter_count..parameter_count + qubit_count)
        .map(|arg_index| {
            Ok(Qubit::Fixed(*match_qis_argument!(
                Qubit,
                arguments,
                arg_index,
                function_name
            )?))
        })
        .collect::<Result<Vec<Qubit>>>()?;

    let mut modifiers = vec![];

    if adjoint {
        modifiers.push(GateModifier::Dagger);
    }

    if controlled {
        modifiers.push(GateModifier::Controlled);
    }

    let instruction = Instruction::Gate(Gate {
        name: name.to_owned(),
        parameters,
        qubits,
        modifiers,
    });

    pattern_context.quil_program.add_instruction(instruction);
    Ok(())
}

static QIS_INTRINSIC_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"^__quantum__qis__(?P<operation>[^_]+)(?P<controlled>__ctl)?(?P<adjoint>__adj)?(__body)?$",
    )
    .unwrap()
});

static RT_RECORD_OUTPUT_INTRINSIC_REGEX: LazyLock<Regex> =
    LazyLock::new(|| Regex::new("^__quantum__rt__(?P<record_type>.+)_record_output$").unwrap());

pub(crate) fn rt_record_instruction<'ctx>(
    context: &QCSCompilerContext<'ctx>,
    pattern_context: &mut UnitaryPatternMatchContext<'ctx>,
    instruction: InstructionValue<'ctx>,
) -> Result<PatternResult<'ctx, ()>> {
    match instruction.get_opcode() {
        inkwell::values::InstructionOpcode::Call => {
            let function_target_name = get_called_function_name(instruction)?;

            if let Some(function_name) = function_target_name {
                if let Some(captures) = RT_RECORD_OUTPUT_INTRINSIC_REGEX.captures(&function_name) {
                    let record_type = &captures["record_type"];

                    match record_type {
                        "result" => {
                            let arguments = get_qis_function_arguments(context, instruction)?;
                            if let Some(OperationArgument::Result(result_index)) = arguments.first()
                            {
                                let next_ro_index =
                                    pattern_context.read_result_mapping.len() as u64;
                                let index = pattern_context.read_result_mapping.entry(*result_index).or_insert_with(|| {
                                    log::info!("Result index {result_index} was read but was never the target of a measurement operation, so recorded output value will always be 0");
                                    next_ro_index
                                });
                                pattern_context
                                    .recorded_output
                                    .push(RecordedOutput::ResultReadoutOffset(*index));
                            } else {
                                return Err(eyre!(
                                    "malformed or missing arguments for: __quantum_rt__{}_record_output",
                                    record_type
                                ));
                            }
                        }
                        "bool" | "integer" | "double" => {
                            return Err(eyre!("unimplemented record type: {}", record_type));
                        }
                        "tuple_start" => pattern_context
                            .recorded_output
                            .push(RecordedOutput::TupleStart),
                        "tuple_end" => pattern_context
                            .recorded_output
                            .push(RecordedOutput::TupleEnd),
                        "array_start" => pattern_context
                            .recorded_output
                            .push(RecordedOutput::ArrayStart),
                        "array_end" => pattern_context
                            .recorded_output
                            .push(RecordedOutput::ArrayEnd),
                        _ => {
                            return Ok(None);
                        }
                    }
                    return Ok(Some((instruction.get_next_instruction(), ())));
                }
            }
            Ok(None)
        }
        _ => Ok(None),
    }
}

#[allow(clippy::too_many_lines)]
pub(crate) fn quantum_instruction<'ctx>(
    context: &QCSCompilerContext<'ctx>,
    pattern_context: &mut UnitaryPatternMatchContext<'ctx>,
    instruction: InstructionValue<'ctx>,
) -> Result<PatternResult<'ctx, ()>> {
    match instruction.get_opcode() {
        inkwell::values::InstructionOpcode::Call => {
            let function_target_name = get_called_function_name(instruction)?;

            if let Some(function_name) = function_target_name {
                if let Some(captures) = QIS_INTRINSIC_REGEX.captures(&function_name) {
                    let operation = &captures["operation"];
                    let adjoint = captures.name("adjoint").is_some();
                    let controlled = captures.name("controlled").is_some();

                    let arguments = get_qis_function_arguments(context, instruction)?;

                    let matched = match operation {
                        "swap" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "SWAP",
                                adjoint,
                                controlled,
                                0,
                                2,
                            )?;
                            true
                        }
                        "toffoli" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "CCNOT",
                                adjoint,
                                controlled,
                                0,
                                3,
                            )?;
                            true
                        }
                        "cnot" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "CNOT",
                                adjoint,
                                controlled,
                                0,
                                2,
                            )?;
                            true
                        }
                        "cz" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "CZ",
                                adjoint,
                                controlled,
                                0,
                                2,
                            )?;
                            true
                        }
                        "h" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "H",
                                adjoint,
                                controlled,
                                0,
                                1,
                            )?;
                            true
                        }
                        "reset" => {
                            // TODO: Alert the user that we're resetting for all qubits instead of just the targeted qubit here
                            pattern_context.use_active_reset = true;
                            true
                        }
                        "rx" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "RX",
                                adjoint,
                                controlled,
                                1,
                                1,
                            )?;
                            true
                        }
                        "ry" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "RY",
                                adjoint,
                                controlled,
                                1,
                                1,
                            )?;
                            true
                        }
                        "rz" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "RZ",
                                adjoint,
                                controlled,
                                1,
                                1,
                            )?;
                            true
                        }
                        "s" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "S",
                                adjoint,
                                controlled,
                                0,
                                1,
                            )?;
                            true
                        }
                        "t" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "T",
                                adjoint,
                                controlled,
                                0,
                                1,
                            )?;
                            true
                        }
                        "x" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "X",
                                adjoint,
                                controlled,
                                0,
                                1,
                            )?;
                            true
                        }
                        "y" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "Y",
                                adjoint,
                                controlled,
                                0,
                                1,
                            )?;
                            true
                        }
                        "z" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "Z",
                                adjoint,
                                controlled,
                                0,
                                1,
                            )?;
                            true
                        }
                        "mz" => {
                            let qubit =
                                *match_qis_argument!(Qubit, arguments, 0, function_name.as_str())?;
                            let result =
                                *match_qis_argument!(Result, arguments, 1, function_name.as_str())?;

                            // Result indices may be sparse rather than increasing monotonically from 0.
                            // If used naively (i.e. %Result 5 as `ro[5]`), this would result in sparse, suboptimal allocation
                            // in the readout data fetched following execution. So, instead, we assign Result indices to
                            // monotonically increasing `ro` region offsets so as to keep `ro` dense.
                            let next_ro_index = pattern_context.read_result_mapping.len() as u64;
                            let ro_buffer_index = pattern_context
                                .read_result_mapping
                                .entry(result)
                                .or_insert_with(|| next_ro_index);

                            pattern_context
                                .quil_program
                                .add_instruction(Instruction::Measurement(Measurement {
                                    target: Some(MemoryReference {
                                        name: String::from("ro"),
                                        index: *ro_buffer_index,
                                    }),
                                    qubit: Qubit::Fixed(qubit),
                                }));

                            true
                        }
                        _ => false,
                    };

                    if matched {
                        pattern_context.instructions_to_remove.push(instruction);
                        Ok(Some((instruction.get_next_instruction(), ())))
                    } else {
                        Ok(None)
                    }
                } else if function_name == "__quantum__qis__read_result__body" {
                    let arguments = get_qis_function_arguments(context, instruction)?;
                    if let Some(OperationArgument::Result(result_index)) = arguments.first() {
                        let ro_index = pattern_context.read_result_mapping.get(result_index).ok_or_else(|| eyre!("Result index {} was never the target of a measurement operation", result_index))?;
                        pattern_context
                            .readout_instruction_mapping
                            .push((*ro_index, instruction));
                    } else {
                        // TODO: Support more read results
                        return Err(eyre!("malformed read_result instrinsic"));
                    }
                    pattern_context.instructions_to_remove.push(instruction);
                    Ok(Some((instruction.get_next_instruction(), ())))
                } else {
                    Ok(None)
                }
            } else {
                Ok(None)
            }
        }
        _ => Ok(None),
    }
}
