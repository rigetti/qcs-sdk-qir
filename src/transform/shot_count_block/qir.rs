// Copyright 2022 Rigetti Computing
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use eyre::{eyre, ContextCompat, Result};
use inkwell::{
    basic_block::BasicBlock,
    values::{AnyValue, FunctionValue, InstructionValue},
};
use log::{debug, info};
use quil_rs::instruction::Vector;

use crate::interop::instruction::{
    get_conditional_branch_else_target, remove_instructions_in_safe_order,
    replace_conditional_branch_target, replace_phi_clauses,
};
use crate::{context::QCSCompilerContext, interop::call, interop::entrypoint::get_entry_function};

use super::pattern::ShotCountPatternMatchContext;
use super::PARAMETER_MEMORY_REGION_NAME;

/// Build and insert an LLVM function which performs up-front translation of
/// all Quil programs used in the module and stores them in an array referred
/// to as the "executable cache".
pub(crate) fn build_populate_executable_cache_function<'ctx>(
    context: &mut QCSCompilerContext<'ctx>,
) -> Result<FunctionValue<'ctx>> {
    let populate_executable_array_function_type =
        context.base_context.void_type().fn_type(&[], false);
    let populate_executable_array_function = context.module.add_function(
        "populate_executable_array",
        populate_executable_array_function_type,
        None,
    );
    let basic_block = context
        .base_context
        .append_basic_block(populate_executable_array_function, "entry");

    context.builder.position_at_end(basic_block);

    let actual_executable_cache = context
        .builder
        .build_call(
            context.values.create_executable_cache(),
            &[context
                .base_context
                .i32_type()
                .const_int(context.quil_programs.len() as u64, false)
                .into()],
            "",
        )
        .try_as_basic_value()
        .left()
        .ok_or_else(|| eyre!("create_executable_cache does not have a return value"))?
        .into_pointer_value();

    context.builder.build_store(
        context.values.executable_cache().as_pointer_value(),
        actual_executable_cache,
    );

    for index in 0..context.quil_programs.len() {
        let program_text = context.quil_programs[index].to_string(true);

        let quil_program_global_string = unsafe {
            // NOTE: this segfaults if the builder is not already positioned within a basic block
            // see https://github.com/TheDan64/inkwell/issues/32
            context
                .builder
                .build_global_string(&program_text, "quil_program")
        };

        context.builder.build_call(
            context.values.add_executable_cache_item(),
            &[
                actual_executable_cache.into(),
                context
                    .base_context
                    .i32_type()
                    .const_int(index as u64, false)
                    .into(),
                quil_program_global_string
                    .as_pointer_value()
                    .const_cast(context.types.string())
                    .into(),
            ],
            "",
        );
    }

    context.builder.build_return(None);

    Ok(populate_executable_array_function)
}

/// Mutate a context such that all contiguous instructions which may be transpiled by `transpile_instruction`
/// are inlined and executed using a shared library call.
#[allow(dead_code)]
pub(crate) fn transpile_module(context: &mut QCSCompilerContext) -> Result<()> {
    let entrypoint_function = get_entry_function(&context.module)
        .ok_or_else(|| eyre!("entrypoint not found in module"))?;
    transpile_function(context, entrypoint_function, &[])?;
    let populate_function = build_populate_executable_cache_function(context)?;

    let entry_basic_block = entrypoint_function
        .get_first_basic_block()
        .ok_or_else(|| eyre!("entrypoint function has no basic blocks"))?;

    match entry_basic_block.get_first_instruction() {
        Some(instruction) => context.builder.position_before(&instruction),
        None => context.builder.position_at_end(entry_basic_block),
    };

    context.builder.build_call(populate_function, &[], "");

    Ok(())
}

pub(crate) fn transpile_function<'ctx>(
    context: &mut QCSCompilerContext<'ctx>,
    function: FunctionValue<'ctx>,
    visited_functions: &[&str],
) -> eyre::Result<()> {
    for current_basic_block in function.get_basic_blocks() {
        transpile_basic_block(context, current_basic_block, visited_functions)?;
    }
    Ok(())
}

pub(crate) fn transpile_basic_block<'ctx>(
    context: &mut QCSCompilerContext<'ctx>,
    basic_block: BasicBlock<'ctx>,
    visited_functions: &[&str],
) -> eyre::Result<()> {
    let pattern_context = ShotCountPatternMatchContext::from_basic_block(
        context,
        basic_block,
        visited_functions,
        transpile_function,
    )?;

    insert_quil_program(context, pattern_context, basic_block)
}

/// Insert the quil program which has been collected from a `BasicBlock` (if any):
///
/// 1. Create a global variable with the program text
/// 2. Insert a shared library call to execute that program text with shot count
/// 3. Remove all of the relevant instructions from the program
#[allow(clippy::too_many_lines, clippy::unnecessary_wraps)]
pub(crate) fn insert_quil_program<'ctx, 'p: 'ctx>(
    context: &mut QCSCompilerContext<'ctx>,
    pattern_context: ShotCountPatternMatchContext<'p>,
    basic_block: BasicBlock,
) -> eyre::Result<()> {
    if let Some((program, shots)) = pattern_context.get_program_data() {
        debug!(
            "inserting quil program with {} shots: {}",
            shots,
            program.to_string(true)
        );

        let mut program = program.clone();

        program.add_instruction(quil_rs::instruction::Instruction::Declaration(
            quil_rs::instruction::Declaration {
                name: String::from("ro"),
                size: Vector {
                    data_type: quil_rs::instruction::ScalarType::Bit,
                    length: pattern_context.read_result_mapping.len() as u64,
                },
                sharing: None,
            },
        ));

        if !pattern_context.get_dynamic_parameters().is_empty() {
            program.add_instruction(quil_rs::instruction::Instruction::Declaration(
                quil_rs::instruction::Declaration {
                    name: String::from(PARAMETER_MEMORY_REGION_NAME),
                    size: Vector {
                        data_type: quil_rs::instruction::ScalarType::Real,
                        length: pattern_context.get_dynamic_parameters().len() as u64,
                    },
                    sharing: None,
                },
            ));
        }

        if pattern_context.use_active_reset {
            // Prepend a reset to the program via copy
            let instructions = program.to_instructions(true);
            let mut new_program = quil_rs::program::Program::new();
            new_program.add_instruction(quil_rs::instruction::Instruction::Reset(
                quil_rs::instruction::Reset { qubit: None },
            ));
            for instruction in instructions {
                new_program.add_instruction(instruction);
            }
            program = new_program;
        }

        if let Some(rewiring_pragma) = &context.options.rewiring_pragma {
            // Prepend a pragma to the program via copy
            let instructions = program.to_instructions(true);
            let mut new_program = quil_rs::program::Program::new();
            new_program.add_instruction(quil_rs::instruction::Instruction::Pragma(
                quil_rs::instruction::Pragma {
                    name: String::from("INITIAL_REWIRING"),
                    arguments: vec![format!("\"{}\"", rewiring_pragma.clone())],
                    data: None,
                },
            ));
            for instruction in instructions {
                new_program.add_instruction(instruction);
            }
            program = new_program;
        }

        // We write all the new instructions to a new basic block
        let execution_basic_block = context.base_context.insert_basic_block_after(
            basic_block,
            &format!("{}_execution", basic_block.get_name().to_str()?),
        );
        basic_block.replace_all_uses_with(&execution_basic_block);
        context.builder.position_at_end(execution_basic_block);

        let executable = if context.options.cache_executables {
            let quil_program_index = context.quil_programs.len();
            context.quil_programs.push(program);

            call::get_executable(
                context,
                context
                    .base_context
                    .i32_type()
                    .const_int(quil_program_index as u64, false),
            )?
        } else {
            let program_text = program.to_string(true);
            let quil_program_global_string = unsafe {
                // NOTE: this segfaults if the builder is not already positioned within a basic block
                // see https://github.com/TheDan64/inkwell/issues/32
                context
                    .builder
                    .build_global_string(&program_text, "quil_program")
            };

            // Insert the shared library calls to send this program for execution
            call::executable_from_quil(context, quil_program_global_string.as_pointer_value())?
        };

        call::wrap_in_shots(context, &executable, shots);

        for (index, value) in pattern_context.parameters.iter().enumerate() {
            call::set_param(context, &executable, index as u64, *value);
        }

        let execution_result = match &context.target {
            crate::context::target::ExecutionTarget::Qpu(_) => {
                call::execute_on_qpu(context, &executable)?
            }
            crate::context::target::ExecutionTarget::Qvm => {
                call::execute_on_qvm(context, &executable)?
            }
        };

        call::panic_on_execution_result_failure(context, &execution_result);

        // After execution we branch into the reduction block, which is everything left over after the
        // quantum instructions are removed.
        context.builder.build_unconditional_branch(basic_block);

        // We place our cursor right after the beginning of the loop, which should come before any reduction instructions.
        context.builder.position_at(
            basic_block,
            &pattern_context
                .initial_instruction
                .and_then(InstructionValue::get_next_instruction)
                .ok_or_else(|| eyre!("Expected an initial instruction"))?,
        );

        let shot_index = pattern_context
            .initial_instruction
            .ok_or_else(|| eyre!("Expected an initial instruction"))?
            .as_any_value_enum()
            .into_int_value();

        for (readout_index, instruction) in &pattern_context.readout_instruction_mapping {
            let new_instruction =
                call::get_readout_bit(context, &execution_result, shot_index, *readout_index)?;

            instruction.replace_all_uses_with(
                &new_instruction
                    .as_instruction()
                    .ok_or_else(|| eyre!("Expected an instruction"))?,
            );
        }

        let cleanup_basic_block = context.base_context.insert_basic_block_after(
            basic_block,
            format!("{}_cleanup", basic_block.get_name().to_str()?).as_str(),
        );

        // Record which block was originally the target following execution & processing of shots in this block
        let original_next_block = get_conditional_branch_else_target(
            basic_block
                .get_terminator()
                .ok_or_else(|| eyre!("Expected a terminator"))?,
        )
        .wrap_err("expected the basic block to have a conditional 'else' target")?;

        context.builder.position_at_end(cleanup_basic_block);
        call::free_execution_result(context, &execution_result);
        context
            .builder
            .build_unconditional_branch(original_next_block);

        replace_conditional_branch_target(
            context,
            basic_block
                .get_terminator()
                .ok_or_else(|| eyre!("Expected a terminator"))?,
            Some(&basic_block),
            Some(&cleanup_basic_block),
        )?;

        replace_phi_clauses(
            context,
            basic_block,
            basic_block,
            execution_basic_block,
            true,
        )?;

        remove_instructions_in_safe_order(pattern_context.instructions_to_remove);

        info!(
            "transpiled basic block {}",
            basic_block.get_name().to_string_lossy()
        );
        info!(
            "inserted basic block {}",
            execution_basic_block.get_name().to_string_lossy()
        );
    } else {
        debug!(
            "not inserting quil program, pattern context: {:?}",
            pattern_context
        );
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;

    mod can_transpile_program_with {
        use crate::context::context::{ContextOptions, QCSCompilerContext};
        use crate::context::target::ExecutionTarget;

        use super::*;

        macro_rules! make_snapshot_test {
            ($name:ident) => {
                #[test]
                fn $name() {
                    let _ = env_logger::builder().is_test(true).try_init();

                    let base_context = inkwell::context::Context::create();
                    let data =
                        std::fs::read(format!("tests/fixtures/programs/{}.bc", stringify!($name)))
                            .unwrap();
                    let mut context = QCSCompilerContext::new_from_data(
                        &base_context,
                        &data,
                        ExecutionTarget::Qvm,
                        ContextOptions {
                            cache_executables: false,
                            rewiring_pragma: None,
                        },
                    )
                    .unwrap();
                    transpile_module(&mut context).expect("transpilation failed");

                    insta::assert_snapshot!(context.module.print_to_string().to_str().unwrap());
                }
            };
        }

        make_snapshot_test!(shot_count_loop);
        make_snapshot_test!(measure);
        make_snapshot_test!(measure_sparse);
        make_snapshot_test!(parametric);
        make_snapshot_test!(reduction);
        make_snapshot_test!(vqe_iteration);
        make_snapshot_test!(cartesian_rotations);
    }
}
