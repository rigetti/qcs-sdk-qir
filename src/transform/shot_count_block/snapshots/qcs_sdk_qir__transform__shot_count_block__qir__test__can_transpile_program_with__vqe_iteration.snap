---
source: src/transform/shot_count_block/qir.rs
expression: context.module.print_to_string().to_str().unwrap()
---
; ModuleID = 'qcs'
source_filename = "./tests/fixtures/programs/vqe_iteration.ll"

%ExecutableCache = type opaque
%Qubit = type opaque
%Result = type opaque
%Executable = type opaque
%ExecutionResult = type opaque

@executable_cache = private global %ExecutableCache* null
@parameter_memory_region_name = private unnamed_addr constant [12 x i8] c"__qir_param\00", align 1
@quil_program = private unnamed_addr constant [138 x i8] c"DECLARE ro BIT[1]\0ARESET\0ACNOT 1 0\0ACNOT 1 0\0AH 0\0AH 1\0AS 1\0AH 1\0ACNOT 1 0\0ACNOT 1 0\0AH 1\0ADAGGER S 1\0AH 1\0AH 0\0AH 2\0ACZ 2 0\0ACZ 2 1\0AH 2\0AMEASURE 2 ro[0]\0A\00", align 1

declare void @__quantum__qis__reset__body(%Qubit*) local_unnamed_addr

declare void @__quantum__qis__s__body(%Qubit*) local_unnamed_addr

declare void @__quantum__qis__s__adj(%Qubit*) local_unnamed_addr

declare void @__quantum__qis__cnot__body(%Qubit*, %Qubit*) local_unnamed_addr

declare void @__quantum__qis__cz__body(%Qubit*, %Qubit*) local_unnamed_addr

declare void @__quantum__qis__h__body(%Qubit*) local_unnamed_addr

declare void @__quantum__qis__rz__body(double, %Qubit*) local_unnamed_addr

; Function Attrs: nofree nosync nounwind readnone speculatable willreturn
declare double @llvm.pow.f64(double, double) #0

declare void @__quantum__qis__mz__body(%Qubit*, %Result*) local_unnamed_addr

declare i1 @__quantum__qis__read_result__body(%Result*) local_unnamed_addr

define internal fastcc void @QuantumApplication__Run__body() unnamed_addr {
entry:
  call void @populate_executable_array()
  br label %body_execution

body:                                             ; preds = %body, %body_execution
  %0 = phi i64 [ %spec.select.i17.i53, %body ], [ 0, %body_execution ]
  %1 = phi i64 [ %4, %body ], [ 1, %body_execution ]
  %2 = call i1 @get_readout_bit(%ExecutionResult* %7, i64 %1, i64 0)
  %3 = zext i1 %2 to i64
  %spec.select.i17.i53 = add i64 %0, %3
  %4 = add nuw nsw i64 %1, 1
  %5 = icmp ult i64 %1, 1000
  br i1 %5, label %body, label %body_cleanup

body_cleanup:                                     ; preds = %body
  call void @free_execution_result(%ExecutionResult* %7)
  br label %exit

body_execution:                                   ; preds = %entry, <null operand!>
  %6 = call %Executable* @executable_from_quil(i8* getelementptr inbounds ([138 x i8], [138 x i8]* @quil_program, i32 0, i32 0))
  call void @wrap_in_shots(%Executable* %6, i32 1000)
  %7 = call %ExecutionResult* @execute_on_qvm(%Executable* %6)
  call void @panic_on_failure(%ExecutionResult* %7)
  br label %body

exit:                                             ; preds = %body_cleanup, <null operand!>
  ret void
}

declare %Executable* @executable_from_quil(i8*)

declare %ExecutionResult* @execute_on_qpu(%Executable*, i8*)

declare %ExecutionResult* @execute_on_qvm(%Executable*)

declare void @free_executable(%Executable*)

declare void @free_execution_result(%ExecutionResult*)

declare i1 @get_readout_bit(%ExecutionResult*, i64, i64)

declare void @panic_on_failure(%ExecutionResult*)

declare void @set_param(%Executable*, i8*, i32, double)

declare void @wrap_in_shots(%Executable*, i32)

declare %ExecutableCache* @create_executable_cache(i32)

declare void @add_executable_cache_item(%ExecutableCache*, i32, i8*)

declare %Executable* @read_from_executable_cache(%ExecutableCache*, i32)

declare void @free_executable_cache(%ExecutableCache*)

define void @populate_executable_array() {
entry:
  %0 = call %ExecutableCache* @create_executable_cache(i32 0)
  store %ExecutableCache* %0, %ExecutableCache** @executable_cache, align 8
  ret void
}

attributes #0 = { nofree nosync nounwind readnone speculatable willreturn }

